### PR TITLE
[FDBSYNCD] Added support for FDBSYNCD for EVPN as described in the PR Azure/SONiC#437

### PR DESCRIPTION
**Pull request by kishorekunal01**
[FDBSYNCD] Added support for FDBSYNCD for EVPN as described in the PR Azure/SONiC#437

This Change set is to support EVPN MAC/IMET routes  from Kernel to DB and DB to Kernel.

**What I did**
1. New demon is added to sync MAC from Kernel to DB and vise versa.
2. Add remote IMET route learned via EVPN to EVPN_REMOTE_VNI_TABLE in APP_DB.
3. Add remote MAC route learned via EVPN to VXLAN_FDB_TABLE.
4. Add State DB FDB_TABLE for local learned MAC to Kernel

**Why I did it**
Part of overall EVPN-VXLAN Implementation

**How I verified it**
Not yet compiled or verified

**Details if related**
HLD Location : https://github.com/Azure/SONiC/pull/437/commits

Signed-off-by: Kishore Kunal <kishore.kunal@broadcom.com>

### CODE REVIEW

**Comment by lguohan** [comment]
need vs test to cover this feature.

File: fdbsyncd/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,17 @@
+INCLUDES = -I $(top_srcdir) -I $(top_srcdir)/warmrestart   -I ../nlapi/include -I ../cfgmgr
```
Comment (gord1306) [inline fdbsyncd/Makefile.am:1]: Hi @kishorekunal01  May I know `nlapi` is related to which PRs, thanks

File: fdbsyncd/fdbsync.cpp [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "ipaddress.h"
+#include "netmsg.h"
+#include "macaddress.h"
+#include "exec.h"
+#include "shellcmd.h"
+#include "fdbsync.h"
+#include "warm_restart.h"
+#include "nlapi.h"
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:18]: Is this `nlapi.h` is related to any other PRs, thanks

File: fdbsyncd/Makefile.am (hunk 1 above)
Comment (kishorekunal01) [reply fdbsyncd/Makefile.am:1]: Yes, nlapi is in another PR. I will remove this from this PR.

File: fdbsyncd/fdbsync.cpp (hunk 2 above)
Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp:18]: I will remove nlapi.h  as that will come in other PR

File: fdbsyncd/fdbsync.cpp [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+#include "errno.h"
+
+using namespace std;
+using namespace swss;
+
+#define VXLAN_BR_IF_NAME_PREFIX    "Brvxlan"
+
+FdbSync::FdbSync(RedisPipeline *pipelineAppDB, DBConnector *stateDb) :
+    m_fdbTable(pipelineAppDB, APP_VXLAN_FDB_TABLE_NAME),
+    m_imetTable(pipelineAppDB, APP_EVPN_REMOTE_VNI_TABLE_NAME),
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:27]: According to sonic-swss-common [PR339](https://github.com/Azure/sonic-swss-common/pull/339), should this constant be renamed to `APP_VXLAN_REMOTE_VNI_TABLE_NAME`

File: fdbsyncd/fdbsync.cpp [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+    m_stateFdbRestoreTable(stateDb, STATE_EVPN_FDB_RESTORE_TABLE_NAME),
+    m_fdbStateTable(stateDb, STATE_FDB_TABLE_NAME)
+{
+    m_AppRestartAssist = new AppRestartAssist(pipelineAppDB, "fdbsyncd", "swss", DEFAULT_FDBSYNC_WARMSTART_TIMER);
+    if (m_AppRestartAssist)
+    {
+        m_AppRestartAssist->registerAppTable(APP_VXLAN_FDB_TABLE_NAME, &m_fdbTable);
+        m_AppRestartAssist->registerAppTable(APP_EVPN_REMOTE_VNI_TABLE_NAME, &m_imetTable);
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:35]: According to sonic-swss-common [PR339](https://github.com/Azure/sonic-swss-common/pull/339), should this constant be renamed to `APP_VXLAN_REMOTE_VNI_TABLE_NAME`

File: fdbsyncd/fdbsync.cpp [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+    std::vector<FieldValueTuple> fvVector;
+    FieldValueTuple f("vni", to_string(vni));
+    fvVector.push_back(f);
+
+    imet_route[key].redist_add_status = true;
+    
+	// If warmstart is in progress, we take all netlink changes into the cache map
+    if (m_AppRestartAssist->isWarmStartInProgress())
+    {
+        m_AppRestartAssist->insertToMap(APP_EVPN_REMOTE_VNI_TABLE_NAME, key, fvVector, false);
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:303]: According to sonic-swss-common [PR339](https://github.com/Azure/sonic-swss-common/pull/339), should this constant be renamed to `APP_VXLAN_REMOTE_VNI_TABLE_NAME`

File: fdbsyncd/fdbsync.cpp [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+			key.c_str(), inet_ntoa(vtep), vlan_id.c_str());
+
+    std::vector<FieldValueTuple> fvVector;
+    FieldValueTuple f("vni", to_string(vni));
+    fvVector.push_back(f);
+
+    // If warmstart is in progress, we take all netlink changes into the cache map
+    if (m_AppRestartAssist->isWarmStartInProgress())
+    {
+        m_AppRestartAssist->insertToMap(APP_EVPN_REMOTE_VNI_TABLE_NAME, key, fvVector, true);
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:330]: According to sonic-swss-common [PR339](https://github.com/Azure/sonic-swss-common/pull/339), should this constant be renamed to `APP_VXLAN_REMOTE_VNI_TABLE_NAME`

File: fdbsyncd/fdbsync.cpp (hunk 3 above)
Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp]: Thanks for review. Updated the same

File: fdbsyncd/fdbsync.cpp (hunk 4 above)
Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp]: Thanks for review. Updated the same

File: fdbsyncd/fdbsync.cpp (hunk 5 above)
Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp]: Thanks for review. Updated the same

File: fdbsyncd/fdbsync.cpp (hunk 6 above)
Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp]: Thanks for review. Updated the same

File: fdbsyncd/fdbsync.cpp [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+        if (macCheckSrcDB(info) == false)
+        {
+            return;
+        }
+		port_name+= fdbmac[key].port_name;
+		fdb_type = fdbmac[key].type;
+		fdbmac.erase(key);
+    }
+
+    if (fdb_type == FDB_TYPE_DYNAMIC)
```
Comment (gord1306) [inline fdbsyncd/fdbsync.cpp:188]: There is a question about the bridge fdb command, whether the remote MAC need to have `extern_learn` field

Comment (kishorekunal01) [reply fdbsyncd/fdbsync.cpp:188]: Thanks for review, Yes extern_learn flag need to be set but currently exec command doesn't has extern_learn support, Hence it is not set.

File: fdbsyncd/fdbsync.cpp [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+        return;
+    }
+
+    if (rtnl_link_vxlan_get_id(link, &vni) != 0)
+    {
+        SWSS_LOG_NOTICE("Op:%d VxLAN dev:%s index:%d vni:%d. Not found", nlmsg_type, ifname? ifname: nil, ifindex, vni);
+        return;
+    }
+    SWSS_LOG_INFO("Op:%d VxLAN dev %s index:%d vni:%d", nlmsg_type, ifname? ifname: nil, ifindex, vni);
+    if (nlmsg_type == RTM_NEWLINK)
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:578]: So we are not handling RTM_DELLINK?

File: fdbsyncd/fdbsync.cpp [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+        isIntfExist = true;
+        ifname = intf_info[ifindex].ifname;
+    }
+
+    nl_addr2str(rtnl_neigh_get_lladdr(neigh), macStr, MAX_ADDR_SIZE);
+
+    if (isIntfExist == false)
+    {
+        /* In Normal run kernel should not ageout this mac */
+        isNotVxlanLink = true;
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:468]: why dont you reuse `isIntfExist `?

File: fdbsyncd/fdbsync.cpp [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+
+        /* When the mac is not learned on the vxlan interface then check if we need to refresh the route */
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:470]: This comment doesn't align with code. Can you rephrase or delete the comment?

File: fdbsyncd/fdbsync.cpp [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+    {
+        isIntfExist = true;
+        ifname = intf_info[ifindex].ifname;
+    }
+
+    nl_addr2str(rtnl_neigh_get_lladdr(neigh), macStr, MAX_ADDR_SIZE);
+
+    if (isIntfExist == false)
+    {
+        /* In Normal run kernel should not ageout this mac */
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:467]: Comment doesn't align with code. Please rephrase or delete

File: fdbsyncd/fdbsync.cpp [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+        isNotVxlanLink = true;
+
+        /* When the mac is not learned on the vxlan interface then check if we need to refresh the route */
+        if (nlmsg_type != RTM_DELNEIGH)
+            return;
+    }
+    else
+    {
+        /* If this is for vnet bridge vxlan interface, then return */
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:476]: If this is *not* for ?

File: fdbsyncd/fdbsync.cpp [hunk 13]
```diff
@@ -0,0 +1,603 @@
...
+        if (ifname.find(VXLAN_BR_IF_NAME_PREFIX) != string::npos)
+            return;
+
+        /* VxLan netdevice should be in <name>-<vlan-id> format */
+        str_loc = ifname.find("-");
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:481]: Why do we have this restriction? Its not mentioned in HLD right?

File: fdbsyncd/fdbsync.cpp [hunk 14]
```diff
@@ -0,0 +1,603 @@
...
+
+    std::string res;
+    int ret = swss::exec(cmds, res);
+
+    SWSS_LOG_NOTICE("cmd:%s, res=%s, ret=%d", cmds.c_str(), res.c_str(), ret);
+
+    return;
+}
+
+void FdbSync::macRefreshStateDB(int vlan, string kmac)
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:209]: Lets discuss this function in WG meeting. Not sure I understand this flow.

File: fdbsyncd/fdbsync.cpp [hunk 15]
```diff
@@ -0,0 +1,603 @@
...
+		{
+			SWSS_LOG_INFO("Failed cmd:%s, res=%s, ret=%d", cmds.c_str(), res.c_str(), ret);
+		}
+
+		SWSS_LOG_NOTICE("Refreshing cmd:%s, res=%s, ret=%d", cmds.c_str(), res.c_str(), ret);
+    }
+    return;
+}
+
+bool FdbSync::checkImetExit(string key, uint32_t vni)
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:253]: `checkImetExist` - Typo, please rename

File: fdbsyncd/fdbsync.h [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        unsigned int type;
+        unsigned int vni;
+        std::string  ifname;
+    };
+    std::unordered_map<std::string, macinfo> m_mac;
+
+    struct imetinfo
+    {
+        unsigned int vni;
+        bool         redist_add_status;
```
Comment (prsunny) [inline fdbsyncd/fdbsync.h:102]: Not sure of the usage of this variable. Looks like not intended?

File: fdbsyncd/fdbsync.cpp [hunk 17]
```diff
@@ -0,0 +1,603 @@
...
+{
+    if (imet_route[key].redist_add_status == true)
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:255]: What if key doesn't exist? It will add an entry. Doesn't appear to be doing its purpose as function name

File: fdbsyncd/fdbsync.cpp [hunk 18]
```diff
@@ -0,0 +1,603 @@
...
+
+void FdbSync::imetDelRoute(struct in_addr vtep, string vlan_str, uint32_t vni)
+{
+    string vlan_id = "Vlan" + vlan_str;
+    string key = vlan_id + ":" + inet_ntoa(vtep);
+
+    if (!checkDelImet(key, vni))
+        return;
+
+	SWSS_LOG_NOTICE("%sIMET Del route key:%s vtep:%s %s",
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:319]: Please align code and also use {} for all `ifs`

File: fdbsyncd/fdbsync.cpp [hunk 19]
```diff
@@ -0,0 +1,603 @@
...
+    {
+        m_AppRestartAssist->insertToMap(APP_VXLAN_REMOTE_VNI_TABLE_NAME, key, fvVector, true);
+        return;
+    }
+    
+	m_imetTable.del(key);
+    return;
+}
+
+void FdbSync::macUpdateVxlanDB(string key)
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:338]: Please rename to `macUpdateRemoteFdb `?

File: fdbsyncd/fdbsync.cpp [hunk 20]
```diff
@@ -0,0 +1,603 @@
...
+
+    std::vector<FieldValueTuple> fvVector;
+    FieldValueTuple rv("remote_vtep", vtep);
+    FieldValueTuple t("type", type);
+    FieldValueTuple v("vni", vni);
+    fvVector.push_back(rv);
+    fvVector.push_back(t);
+    fvVector.push_back(v);
+
+	SWSS_LOG_NOTICE("%sVXLAN_FDB_TABLE: ADD_KEY %s vtep:%s type:%s",
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:357]: Change to INFO for all frequent events that happens like FDB learning etc. There is ratelimiting for syslogs and we will start losing relevent messages

File: fdbsyncd/fdbsync.cpp [hunk 21]
```diff
@@ -0,0 +1,603 @@
...
+    if ((nlmsg_type == RTM_DELNEIGH) || (state == NUD_INCOMPLETE) ||
+        (state == NUD_FAILED))
+    {
+        delete_key = true;
+    }
+
+    if (state & NUD_NOARP)
+    {
+        /* This is a static route */
+        type = true;
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:519]: type is defined as uint32 but used as bool. Do you want to use string itself to mention "static"

File: fdbsyncd/fdbsync.cpp [hunk 22]
```diff
@@ -0,0 +1,603 @@
...
+    
+	SWSS_LOG_NOTICE("VXLAN_FDB_TABLE: DEL_KEY %s vtep:%s type:%s", key.c_str(), vtep.c_str(), type.c_str());
+    m_fdbTable.del(key);
+    return;
+
+}
+
+void FdbSync::macAddVxlan(string key, struct in_addr vtep, uint32_t type, uint32_t vni, string intf_name)
+{
+    m_mac[key].vtep   =  inet_ntoa(vtep);
```
Comment (prsunny) [inline fdbsyncd/fdbsync.cpp:410]: You could optimize it by `m_mac[key] = { vtep, type, vni, intf_name }` and use the same function to write to DB so as to avoid number of functions.

**Review by prsunny — CHANGES_REQUESTED** [review]

File: fdbsyncd/fdbsync.cpp [hunk 23]
```diff
@@ -0,0 +1,603 @@
...
+    m_stateFdbRestoreTable.hget("Flags", "restored", value);
+    if (value == "true")
+    {
+        SWSS_LOG_NOTICE("neighbor table restore to kernel is done");
+        return true;
+    }
+    return true;
+}
+
+void FdbSync::processStateFdb()
```
[the remark this answers is not in the document]
Comment (prsunny) [reply fdbsyncd/fdbsync.cpp:61]: Lets have this fdb programming only if EVPN NVO configuration is present.

**Review by kishorekunal01 — COMMENTED** [review]
All review comment address. WIll raise 2 more PR for the sonic-buildimage to start fdbsyncd process, and adding a table for warm reboot store in swss-common

**Comment by prsunny** [comment]
retest this please

**Review by prsunny — APPROVED** [review]
